### PR TITLE
[fix](case expr) fix coredump of case for null value 2

### PR DESCRIPTION
**Pull request by xinyiZzz**
## Proposed changes

Issue Number: close #xxx

more complete fix: #20564

```
F0609 13:58:40.029528 1179830 assert_cast.h:46] Bad cast from type:doris::vectorized::ColumnNullable* to doris::vectorized::ColumnVector<unsigned char> const*
*** Check failure stack trace: ***
    @     0x5650c8670e2d  google::LogMessage::Fail()
    @     0x5650c8673369  google::LogMessage::SendToLog()
    @     0x5650c8670996  google::LogMessage::Flush()
    @     0x5650c86739d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x5650c1f97c04  assert_cast<>()
    @     0x5650c52b39e4  doris::vectorized::FunctionCase<>::execute_impl<>()
    @     0x5650c52b260d  doris::vectorized::FunctionCase<>::execute_get_then_null<>()
    @     0x5650c52a4b97  doris::vectorized::FunctionCase<>::execute_get_when_null<>()
    @     0x5650c52a3d97  doris::vectorized::FunctionCase<>::execute_get_type()
    @     0x5650c52a3a5c  doris::vectorized::FunctionCase<>::execute_impl()
    @     0x5650c3ea2284  doris::vectorized::DefaultExecutable::execute_impl()
    @     0x5650c4e6187f  doris::vectorized::PreparedFunctionImpl::_execute_skipped_constant_deal()
    @     0x5650c4e5eda5  doris::vectorized::PreparedFunctionImpl::execute_without_low_cardinality_columns()
    @     0x5650c4e5ee20  doris::vectorized::PreparedFunctionImpl::execute()
    @     0x5650c3ea0d3e  doris::vectorized::IFunctionBase::execute()
    @     0x5650c3ece245  doris::vectorized::VCaseExpr::execute()
    @     0x5650c3ed0230  doris::vectorized::VExprContext::execute()
    @     0x5650c3e6e5a7  doris::vectorized::AggFnEvaluator::_calc_argment_columns()
    @     0x5650c3e6e2ed  doris::vectorized::AggFnEvaluator::execute_single_add()
    @     0x5650c3b797c0  doris::vectorized::AggregationNode::_execute_without_key()
    @     0x5650c3c968a7  std::__invoke_impl<>()
    @     0x5650c3c96780  _ZSt10__invoke_rIN5doris6StatusERMNS0_10vectorized15AggregationNodeEFS1_PNS2_5BlockEEJRPS3_S5_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESC_E4typeEOSD_DpOSE_
    @     0x5650c3c96715  _ZNSt12_Bind_resultIN5doris6StatusEFMNS0_10vectorized15AggregationNodeEFS1_PNS2_5BlockEEPS3_St12_PlaceholderILi1EEEE6__callIS1_JOS5_EJLm0ELm1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE
    @     0x5650c3c96695  std::_Bind_result<>::operator()<>()
    @     0x5650c3c96638  std::__invoke_impl<>()
    @     0x5650c3c965d8  _ZSt10__invoke_rIN5doris6StatusERSt12_Bind_resultIS1_FMNS0_10vectorized15AggregationNodeEFS1_PNS3_5BlockEEPS4_St12_PlaceholderILi1EEEEJS6_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESG_E4typeEOSH_DpOSI_
    @     0x5650c3c964a8  std::_Function_handler<>::_M_invoke()
    @     0x5650c3c180f0  std::function<>::operator()()
    @     0x5650c3b7d2ca  doris::vectorized::AggregationNode::sink()
    @     0x5650c3b7d10d  doris::vectorized::AggregationNode::open()
    @     0x5650bd7008a7  doris::PlanFragmentExecutor::open_vectorized_internal()
    @     0x5650bd700285  doris::PlanFragmentExecutor::open()
*** Query id: b818173876041bb-98411315b230fcca ***
*** Aborted at 1686290320 (unix time) try "date -d @1686290320" if you are using GNU date ***
*** Current BE git commitID: 238c1b64e5 ***
*** SIGABRT unknown detail explain (@0x4200011fed5) received by PID 1179349 (TID 1179830 OR 0x7f2546fbf700) from PID 1179349; stack trace: ***
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at doris/core/be/src/common/signal_handler.h:413
 1# 0x00007F264E8C0400 in /lib64/libc.so.6
 2# __GI_raise in /lib64/libc.so.6
 3# __GI_abort in /lib64/libc.so.6
 4# 0x00005650C867B819 in doris/core/output_run/be/lib/doris_be
 5# 0x00005650C8670E2D in doris/core/output_run/be/lib/doris_be
 6# google::LogMessage::SendToLog() in doris/core/output_run/be/lib/doris_be
 7# google::LogMessage::Flush() in doris/core/output_run/be/lib/doris_be
 8# google::LogMessageFatal::~LogMessageFatal() in doris/core/output_run/be/lib/doris_be
 9# doris::vectorized::ColumnVector<unsigned char> const* assert_cast<doris::vectorized::ColumnVector<unsigned char> const*, doris::vectorized::IColumn const*>(doris::vectorized::IColumn const*&&) in doris/core/output_run/be/lib/doris_be
10# doris::Status doris::vectorized::FunctionCase<false, false>::execute_impl<doris::vectorized::ColumnVector<long>, false, true>(std::shared_ptr<doris::vectorized::IDataType const> const&, doris::vectorized::Block&, unsigned long, doris::vectorized::CaseWhenColumnHolder) at doris/core/be/src/vec/functions/function_case.h:245
11# doris::Status doris::vectorized::FunctionCase<false, false>::execute_get_then_null<doris::vectorized::ColumnVector<long>, false>(std::shared_ptr<doris::vectorized::IDataType const> const&, doris::vectorized::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long) at doris/core/be/src/vec/functions/function_case.h:355
12# doris::Status doris::vectorized::FunctionCase<false, false>::execute_get_when_null<doris::vectorized::ColumnVector<long> >(std::shared_ptr<doris::vectorized::IDataType const> const&, doris::vectorized::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long) in doris/core/output_run/be/lib/doris_be
13# doris::vectorized::FunctionCase<false, false>::execute_get_type(std::shared_ptr<doris::vectorized::IDataType const> const&, doris::vectorized::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long) at doris/core/be/src/vec/functions/function_case.h:408
14# doris::vectorized::FunctionCase<false, false>::execute_impl(doris::FunctionContext*, doris::vectorized::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long) at doris/core/be/src/vec/functions/function_case.h:415
...skipping...
pp:268
19# doris::vectorized::IFunctionBase::execute(doris::FunctionContext*, doris::vectorized::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long, bool) at doris/core/be/src/vec/functions/function.h:171
20# doris::vectorized::VCaseExpr::execute(doris::vectorized::VExprContext*, doris::vectorized::Block*, int*) at doris/core/be/src/vec/exprs/vcase_expr.cpp:113
21# doris::vectorized::VExprContext::execute(doris::vectorized::Block*, int*) at doris/core/be/src/vec/exprs/vexpr_context.cpp:61
22# doris::vectorized::AggFnEvaluator::_calc_argment_columns(doris::vectorized::Block*) at doris/core/be/src/vec/exprs/vectorized_agg_fn.cpp:286
23# doris::vectorized::AggFnEvaluator::execute_single_add(doris::vectorized::Block*, char*, doris::vectorized::Arena*) at doris/core/be/src/vec/exprs/vectorized_agg_fn.cpp:210
24# doris::vectorized::AggregationNode::_execute_without_key(doris::vectorized::Block*) at doris/core/be/src/vec/exec/vaggregation_node.cpp:755
25# doris::Status std::__invoke_impl<doris::Status, doris::Status (doris::vectorized::AggregationNode::*&)(doris::vectorized::Block*), doris::vectorized::AggregationNode*&, doris::vectorized::Block*>(std::__invoke_memfun_deref, doris::Status (doris::vectorized::AggregationNode::*&)(doris::vectorized::Block*), doris::vectorized::AggregationNode*&, doris::vectorized::Block*&&) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:74
26# std::enable_if<is_invocable_r_v<doris::Status, doris::Status (doris::vectorized::AggregationNode::*&)(doris::vectorized::Block*), doris::vectorized::AggregationNode*&, doris::vectorized::Block*>, doris::Status>::type std::__invoke_r<doris::Status, doris::Status (doris::vectorized::AggregationNode::*&)(doris::vectorized::Block*), doris::vectorized::AggregationNode*&, doris::vectorized::Block*>(doris::Status (doris::vectorized::AggregationNode::*&)(doris::vectorized::Block*), doris::vectorized::AggregationNode*&, doris::vectorized::Block*&&) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:114
27# doris::Status std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>::__call<doris::Status, doris::vectorized::Block*&&, 0ul, 1ul>(std::tuple<doris::vectorized::Block*&&>&&, std::_Index_tuple<0ul, 1ul>) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:570
28# doris::Status std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>::operator()<doris::vectorized::Block*>(doris::vectorized::Block*&&) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:629
29# doris::Status std::__invoke_impl<doris::Status, std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>&, doris::vectorized::Block*>(std::__invoke_other, std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>&, doris::vectorized::Block*&&) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
30# std::enable_if<is_invocable_r_v<doris::Status, std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>&, doris::vectorized::Block*>, doris::Status>::type std::__invoke_r<doris::Status, std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>&, doris::vectorized::Block*>(std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)>&, doris::vectorized::Block*&&) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:114
31# std::_Function_handler<doris::Status (doris::vectorized::Block*), std::_Bind_result<doris::Status, doris::Status (doris::vectorized::AggregationNode::*(doris::vectorized::AggregationNode*, std::_Placeholder<1>))(doris::vectorized::Block*)> >::_M_invoke(std::_Any_data const&, doris::vectorized::Block*&&) at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
32# std::function<doris::Status (doris::vectorized::Block*)>::operator()(doris::vectorized::Block*) const at doris/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
33# doris::vectorized::AggregationNode::sink(doris::RuntimeState*, doris::vectorized::Block*, bool) at doris/core/be/src/vec/exec/vaggregation_node.cpp:588
34# doris::vectorized::AggregationNode::open(doris::RuntimeState*) at doris/core/be/src/vec/exec/vaggregation_node.cpp:530
```

case
```
SELECT SUM(CASE (((NULL BETWEEN NULL AND NULL))and(CAST(0.4716 AS BOOLEAN) ))  WHEN ((CAST('-1530390546' AS VARCHAR) ) LIKE ('-1678299490')) THEN (- (+ t0.c0))  WHEN CASE (NULL IN (NULL))  WHEN false THEN (t0.c0 NOT BETWEEN t0.c0 AND 1916517711) END  THEN ((((t0.c0)*(1309461808)))/((- -267268292))) END ) FROM t0;
```

```
CREATE TABLE `t0` (
  `c0` tinyint(4) NULL
) ENGINE=OLAP
DUPLICATE KEY(`c0`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`c0`) BUCKETS 20
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"storage_format" = "V2",
"light_schema_change" = "true",
"disable_auto_compaction" = "false",
"enable_single_replica_compaction" = "false"
); 
1 row in set (0.00 sec)
```

```
mysql [database20]>select * from t0;
+------+
| c0   |
+------+
|    1 |
|    1 |
|    1 |
|    0 |
|   -1 |
+------+
5 rows in set (0.02 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_case.h [hunk 1]
```diff
@@ -332,13 +332,15 @@ class FunctionCase : public IFunction {
         bool then_null = false;
         for (int i = 1 + has_case; i < arguments.size() - has_else; i += 2) {
             auto then_column_ptr = block.get_by_position(arguments[i]).column;
-            if (then_column_ptr->is_nullable() || then_column_ptr->only_null()) {
+            if (then_column_ptr->is_nullable() ||
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_case.h:335]: maybe we can remove `then_column_ptr->is_nullable()`?

**Comment by BiteTheDDDDt** [comment]
better add test case

File: be/src/vec/functions/function_case.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/vec/functions/function_case.h:335]: It seems to be ok, for `ColumnConst`, it seems that `get_data_type() == TypeIndex::Nullable` is the real `is_nullable`

**Comment by xinyiZzz** [comment]
> better add test case

done

**Comment by xinyiZzz** [comment]
run buildall

**Review by BiteTheDDDDt — APPROVED** [review]
LGTM

**Comment by jacktengg** [comment]
LGTM

**Pull request merged**